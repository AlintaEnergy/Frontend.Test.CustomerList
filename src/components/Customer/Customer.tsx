import * as React from "react";
import { Dispatch } from "redux";
import { useDispatch } from "react-redux";
import { ICustomer } from "../../types/types";
import {
  StyledCustomer,
  StyledCustomerDelete,
  StyledCustomerInfo,
  StyledCustomerName,
} from "./StyledCustomer";
import { Toaster } from "react-hot-toast";

type Props = {
  customer: ICustomer;
  removeCustomer: (customer: ICustomer) => void;
};

export const Customer = ({ customer, removeCustomer }: Props) => {
  const dispatch: Dispatch<any> = useDispatch();

  const deleteCustomer = React.useCallback(
    (customer: ICustomer) => {
      dispatch(removeCustomer(customer));
    },
    [dispatch]
  );

  return (
    <>
      <StyledCustomer>
        <StyledCustomerName>
          {customer.firstName} {customer.lastName}
        </StyledCustomerName>
        <StyledCustomerInfo>Email: {customer.email}</StyledCustomerInfo>
        <StyledCustomerInfo>Birthday: {customer.birthDate}</StyledCustomerInfo>
        <StyledCustomerDelete onClick={() => deleteCustomer(customer)}>
          Delete
        </StyledCustomerDelete>
      </StyledCustomer>
      <Toaster />
    </>
  );
};
