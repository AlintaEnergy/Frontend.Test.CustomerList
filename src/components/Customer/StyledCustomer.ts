import styled from "styled-components";

export const StyledCustomer = styled.div`
  box-sizing: border-box;
  min-width: 100%;
  border: 1px solid #ccc;
  margin: 1rem;
  padding: 1rem;
`;

export const StyledCustomerName = styled.h2`
  margin: 0 0 0.5rem;
  padding: 0;
`;

export const StyledCustomerInfo = styled.p`
  margin: 0 0 0.5rem;
  padding: 0;
`;

export const StyledCustomerDelete = styled.button`
  padding: 1rem;
  background-color: #dc1616;
  color: white;
  border: none;
  border-radius: 4px;
  font-size: 16px;
`;

export const TableData = styled.td`
  font-weight: 500;
  font-size:20px;
  border: 1px solid LightGrey;
  height: 100%;
  text-align: center;
  vertical-align: middle;
`;
