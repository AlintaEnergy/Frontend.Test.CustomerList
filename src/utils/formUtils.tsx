// eslint-disable-next-line no-useless-escape
export const nameRegex = /^[A-Za-z\s\'-]*$/;

// Accepting 04xx xxx xxx, (0x) xxxx xxxx, +61 xxx xxx xxx and without space
// eslint-disable-next-line no-useless-escape
export const phoneRegex = /^(?:\+?(61))? ?(?:\((?=.*\)))?(0?[2-57-8])\)? ?(\d\d(?:[- ](?=\d{3})|(?!\d\d[- ]?\d[- ]))\d\d[- ]?\d[- ]?\d{3})$/;

export const toHeaderCase = (input: string) => {
  const replaceInput = input.replace(/([A-Z])/g, " $1");
  const changedInput =
    replaceInput.charAt(0).toUpperCase() + replaceInput.slice(1);
  return changedInput;
};
