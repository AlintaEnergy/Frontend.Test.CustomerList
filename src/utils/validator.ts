export const validateEmptyField = (value:string, errorMessage: string) => {
    return !value ? errorMessage : undefined 
  }

export const validatePattern = (value: string, pattern: RegExp, errorMessage: string) => {
    return !pattern.test(value) ? errorMessage : undefined
  }