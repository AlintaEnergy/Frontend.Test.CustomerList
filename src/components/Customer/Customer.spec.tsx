import userEvent from "@testing-library/user-event";
import { render, screen, waitFor } from "../../utils/testUtils";
import { Customer } from "./Customer";
import { v4 as uuidv4 } from 'uuid';

const customer = {
  id: uuidv4(),
  firstName: "Test",
  lastName: "Tester",
  phoneNumber: "00000000",
};

describe("<Customer />", () => {
  const deleteCustomer = jest.fn();
  
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render a <Customer />", () => {
    const wrapper = render(
      <Customer customer={customer} deleteCustomer={deleteCustomer} />
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render the customer details that matches mock customer's details", () => {
    render(<Customer customer={customer} deleteCustomer={deleteCustomer} />);
    expect(screen.getByRole('heading')).toHaveTextContent('Test Tester');
    expect(screen.getByText(/Phone number: /)).toHaveTextContent("Phone number: 00000000");
  });

  it("should render a delete button", () => {
    render(<Customer customer={customer} deleteCustomer={deleteCustomer} />);
    expect(screen.getByRole("button", {
      name:"Delete"
    })).toBeVisible();
  });

  it("should call deleteCustomer when delete is clicked", async () => {
    render(<Customer customer={customer} deleteCustomer={deleteCustomer} />);
    userEvent.click(screen.getByRole("button"));
    await waitFor(() => {
      expect(deleteCustomer).toHaveBeenCalledWith(customer);
    });
  })
});
