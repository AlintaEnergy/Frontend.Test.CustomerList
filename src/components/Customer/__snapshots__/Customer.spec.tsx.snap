// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Customer /> should render a <Customer /> 1`] = `
<div>
  <div
    class="sc-aXZVg bPyiSZ"
  >
    <h2
      class="sc-gEvEer eIdRVW"
    >
      Test
       
      Tester
    </h2>
    <p
      class="sc-eqUAAy bEjATU"
    >
      Phone number: 
      00000000
    </p>
    <button
      class="sc-fqkvVR dAvLEF"
    >
      Delete
    </button>
  </div>
</div>
`;
